### PR TITLE
[IE CLDNN] Extended eltwise fusing

### PR DESCRIPTION
**Pull request by ahnyoung-paul**
### Details:
 - *Add checking multiple eltwise fusibility rule in fuse_eltwise_f in prepare_primitive_fusing*
 - *Modify JitConstant generation code in Jitter to support multiple eltwise fusing*
 - *Add the code to propagate fused node dependency data to the Jitter*
 - *Add cldnn unit-test for extended eltwise fusing*
 - *Update primitive_db_gen.py to append undefs for include CL file*

### Tickets:
 - 48364

### CODE REVIEW

File: inference-engine/thirdparty/clDNN/kernel_selector/core/common/jitter.cpp [hunk 1]
```diff
@@ -1476,17 +1476,19 @@ JitConstants FusedOpsCodeGenerator::MakeLoadJitConstants(const FusedOpsConfigura
...
     auto idx = conf.bfzyx_idx_order;
     std::string shuffle_var = conf.shuffle_var_name;
     bool is_shuffled = false;
+    auto& fused_op_ids = desc.fused_op_ids;
+    std::vector<std::string> input_vars;
 
-    out_var = GetOutputVarName(in_var);
-    out_type = desc.output_tensor.GetDType();
+    out_var = GetOutputVarName(in_var, desc.op_id);
+    auto out_type = desc.output_tensor.GetDType();
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/common/jitter.cpp:1491]: better to use as const auto&

Comment (ahnyoung-paul) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/common/jitter.cpp]: Fixed it. thanks

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp [hunk 2]
```diff
@@ -811,9 +863,81 @@ void prepare_primitive_fusing::fuse_simple_primitives(program_impl &p) {
...
+            //  follow the legacy checking rule
+            if (fused_node->is_type<convolution>() && fused_node->get_users().size() > 1) {
+                // Allowed new pattern: Eltw1, Act, Eltw2, Elt3, Elt4 are fused to Conv1
+                // * Conv1 -> Eltw1(Add) -> Act(Clamp) -> Eltw2(Mul) -> Eltw3(Div) -> Eltw4(Add) -> Conv2
+                // *   \–----------------------------------->/                          \---------> Eltw5(Div)
+                //
+                // Extended eltwise fusiblity checking rules
+                //
+                // 1. All fusing nodes should be eltwise or activation node
+                // 2. All intermediate fusing nodes except last fusing node(i.e. Eltw4) should have only one user.
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp:877]: Is this true? I think intermediate fusing nodes can have multiple children as  long as they can be merged to one within some depth.. (max_level)

Comment (ahnyoung-paul) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp:877]: it is not true. I  will fix it.

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp [hunk 3]
```diff
@@ -811,9 +863,81 @@ void prepare_primitive_fusing::fuse_simple_primitives(program_impl &p) {
...
+                        auto iter = std::find(node_history.begin(), node_history.end(), user);
+                        if (iter == node_history.end())
+                            push_node_queue(user, current_node.second+1);
+                    }
+                } while (node_queue.size() > 1);
+            } else {
+                merge_allowed = fused_node->get_users().size() == 1;
+            }
 
+            // This fusing can be extended to support peer node in any layout
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp:940]: what do you mean with "any layout"? do you mean graph pattern?

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp [hunk 4]
```diff
@@ -811,9 +863,81 @@ void prepare_primitive_fusing::fuse_simple_primitives(program_impl &p) {
...
                 return;
 
-            // This fusing can be extended to support peer node in any layout
-            bool merge_allowed = fused_node->get_users().size() == 1;
+            bool merge_allowed = true;
+            // If fused node is not convolution and fused node has multiple users,
+            //  follow the legacy checking rule
+            if (fused_node->is_type<convolution>() && fused_node->get_users().size() > 1) {
+                // Allowed new pattern: Eltw1, Act, Eltw2, Elt3, Elt4 are fused to Conv1
+                // * Conv1 -> Eltw1(Add) -> Act(Clamp) -> Eltw2(Mul) -> Eltw3(Div) -> Eltw4(Add) -> Conv2
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp:871]: Currently eltwise is only fused if it is mul or sum right? Can it fuse div too?

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp (hunk 2 above)
Comment (yeonbok) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp:877]: Also please mention like this: 
-  Currently eltwise and activations are allowed to be fused from multiple branches, but technically other fusable operations can be allowed too in the future.

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp (hunk 4 above)
Comment (ahnyoung-paul) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp:871]: div means mul in eltwise name. I just copy yolov5 networks node name but I thinks it makes misunderstanding. I will change it.

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp (hunk 3 above)
Comment (ahnyoung-paul) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp:940]: No, It's the oldest comments before implementing extended eltwise fusing. I will remove it

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp (hunk 2 above)
Comment (ahnyoung-paul) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp]: I upload update commit. thanks

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp (hunk 3 above)
Comment (ahnyoung-paul) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp]: I upload update commit. thanks

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp (hunk 4 above)
Comment (ahnyoung-paul) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp]: I upload update commit. thanks

File: inference-engine/thirdparty/clDNN/kernel_selector/core/kernel_base.cpp [hunk 5]
```diff
@@ -115,17 +115,15 @@ JitConstants KernelBase::MakeFusedOpsJitConstants(const kernel_selector::base_pa
...
             for (size_t i = 0; i < params.fused_ops.size(); i++) {
                 auto fused_dep_codegen = FusedOpsCodeGenerator(params.fused_ops[i]);
-                std::string out_var;
-                Datatype out_type;
                 jit.Merge(fused_dep_codegen.MakeLoadJitConstants(c, params.output));
-                jit.Merge(fused_dep_codegen.MakeOpJitConstants(c, in_name, in_type, out_var, out_type));
-                in_name = out_var;
-                in_type = out_type;
+                jit.Merge(fused_dep_codegen.MakeOpJitConstants(c, in_name, in_type, out_name));
+                // in_name = out_name;
```
Comment (yeonbok) [inline inference-engine/thirdparty/clDNN/kernel_selector/core/kernel_base.cpp:126]: please remove this

Comment (ahnyoung-paul) [reply inference-engine/thirdparty/clDNN/kernel_selector/core/kernel_base.cpp]: Fixed it

**Comment by ahnyoung-paul** [comment]
@vladimir-paramuzov Could please review this PR? Thanks.

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp [hunk 6]
```diff
@@ -811,8 +863,84 @@ void prepare_primitive_fusing::fuse_simple_primitives(program_impl &p) {
...
+                        });
+                        if (iter == node_queue.end()) {
+                            node_queue.push_back(std::make_pair(in_node, level));
+                        }
+                    };
+
+                    // If the any user node is not eltwise(mul / add mode) and activation,
+                    // the current node will be considered as last node and put it back into the node_queue
+                    auto curr_users = current_node.first->get_users();
+                    auto invliad_user_iter = std::find_if(curr_users.begin(), curr_users.end(), [&](cldnn::program_node* user){
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp]: It's typos. I fixed it. thanks.

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp [hunk 7]
```diff
@@ -811,8 +863,84 @@ void prepare_primitive_fusing::fuse_simple_primitives(program_impl &p) {
...
+                        auto ret = (user->is_output() ||
+                                    (!(user->is_type<eltwise>() && user->get_primitive()->input.size() == 2 &&
+                                        (std::find(supported_modes.begin(), supported_modes.end(),
+                                        reinterpret_cast<cldnn::eltwise*>(user)->mode) == supported_modes.end())) &&
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp]: Fixed it. thanks.

File: inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp [hunk 8]
```diff
@@ -497,11 +501,59 @@ void prepare_primitive_fusing::fuse_simple_primitives(program_impl &p) {
...
-            if (input_data.get_users().size() != 1 || activation_node.get_dependencies().size() >= 3)
+            if (activation_node.get_dependencies().size() >= 3)
                 return;
 
+            if (input_data.get_users().size() != 1) {
+                // If input_data has fused primitives,
+                // find original dependency of activation_node using fusing_history
+                // and check the number of users of it.
+                // If the node has multiple users it's not fusible.
+                if (input_data.has_fused_primitives()) {
```
[the remark this answers is not in the document]
Comment (ahnyoung-paul) [reply inference-engine/thirdparty/clDNN/src/graph_optimizer/prepare_primitive_fusing.cpp:523]: Yes, quantize is also need similar checks. I extract to separate function for future reuse. Please review it again. thanks.

**Comment by ahnyoung-paul** [comment]
@vladimir-paramuzov Could you please review it again?
thanks.

**Review by vladimir-paramuzov — APPROVED** [review]

**Pull request merged**